homeassistant/components/eufylife_ble/sensor.py

```diff
@@ -0,0 +1,215 @@
+            connections={(dr.CONNECTION_BLUETOOTH, data.address)},
+        )
+
+    @property
+    def available(self) -> bool:
+        """Determine if the entity is available."""
+        if self._data.client.advertisement_data_contains_state:
+            # If the device only uses advertisement data, just check if the address is present.
+            return async_address_present(self.hass, self._data.address)
+
+        # If the device needs an active connection, availability is based on whether it is connected.
+        return self._data.client.is_connected
+
+    @callback
+    def _handle_state_update(self, *args: Any) -> None:
+        """Handle state update."""
+        self.async_write_ha_state()
+
+    async def async_added_to_hass(self) -> None:
+        """Register callback."""
+        self.async_on_remove(
+            self._data.client.register_callback(self._handle_state_update)
+        )
+
+
+class EufyLifeRealTimeWeightSensorEntity(EufyLifeSensorEntity):
+    """Representation of an EufyLife real-time weight sensor."""
+
+    _attr_name = "Real-time weight"
+    _attr_native_unit_of_measurement = UnitOfMass.KILOGRAMS
+    _attr_device_class = SensorDeviceClass.WEIGHT
+
+    def __init__(self, data: EufyLifeData) -> None:
+        """Initialize the real-time weight sensor entity."""
+        super().__init__(data)
+        self._attr_unique_id = f"{data.address}_real_time_weight"
+
+    @property
+    def native_value(self) -> float | None:
+        """Return the native value."""
+        if self._data.client.state is not None:
+            return self._data.client.state.weight_kg
+        return None
+
+    @property
+    def suggested_unit_of_measurement(self) -> str | None:
+        """Set the suggested unit based on the unit system."""
+        if self.hass.config.units is US_CUSTOMARY_SYSTEM:
+            return UnitOfMass.POUNDS
+
+        return UnitOfMass.KILOGRAMS
+
+
+class EufyLifeWeightSensorEntity(RestoreEntity, EufyLifeSensorEntity):
+    """Representation of an EufyLife weight sensor."""
+
+    _attr_name = "Weight"
+    _attr_native_unit_of_measurement = UnitOfMass.KILOGRAMS
+    _attr_device_class = SensorDeviceClass.WEIGHT
+
+    _weight_kg: float | None = None
+
+    def __init__(self, data: EufyLifeData) -> None:
+        """Initialize the weight sensor entity."""
+        super().__init__(data)
+        self._attr_unique_id = f"{data.address}_weight"
+
+    @property
+    def available(self) -> bool:
+        """Determine if the entity is available."""
+        return True
+
+    @property
+    def native_value(self) -> float | None:
+        """Return the native value."""
+        return self._weight_kg
+
+    @property
+    def suggested_unit_of_measurement(self) -> str | None:
+        """Set the suggested unit based on the unit system."""
+        if self.hass.config.units is US_CUSTOMARY_SYSTEM:
+            return UnitOfMass.POUNDS
+
+        return UnitOfMass.KILOGRAMS
+
```

Comment:
The docs say `suggested_unit_of_measurement` its to `This can be used by integrations to override automatic unit conversion rules`. Is this needed?

Reply:
There aren't any automatic unit conversion rules for the `weight` device class. So if I didn't override this, the entities would use kilograms regardless of the user's configured unit system. This wouldn't be desirable because users using the US customary system would typically expect body weight to be measured in pounds.

Reply:
I need to dig into how the unit conversions all work, but it does look like there is a mapping here:

https://github.com/home-assistant/core/blob/26c419705e30a752394ffb0b00f0d899ef4edb73/homeassistant/components/sensor/const.py#L433

Reply:
I believe that mapping is what allows the user to manually override the unit in the UI.

The automatic unit conversions for the US customary system are defined here (note there is no conversion from kilograms to pounds):
https://github.com/home-assistant/core/blob/dev/homeassistant/util/unit_system.py#L301-L335

Reply:
Thanks. That helped. After digging through the sensor platform I see how it works. There are a LOT of cases here so sadly a bit more complex than I would have hoped.  

https://github.com/home-assistant/core/blob/26c419705e30a752394ffb0b00f0d899ef4edb73/homeassistant/components/sensor/__init__.py#L383

https://github.com/home-assistant/core/blob/26c419705e30a752394ffb0b00f0d899ef4edb73/homeassistant/components/sensor/__init__.py#L541

Your logic makes sense.  👍